### PR TITLE
Migrated leader election to lease API

### LINKED ISSUE #111: Migrate leader election to use Lease API
Current leader election is using endpoints

https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/853cdd7cba6e20753071cdb3e07fe37a6f2fc5a3/controller/controller.go#L867

However, that has some problems and is recommended to use the Lease API instead

https://github.com/kubernetes/kubernetes/issues/80289

### PR DESCRIPTION
**Pull request by NikhilSharmaWe**
Fixes #111

### CODE REVIEW

**Comment by NikhilSharmaWe** [comment]
@aojea Why is the change here in wrong. I asked wojtekt (on slack) about how to approach this issue, he said that we have to just reuse the existing infrastructure now. So he suggested to change `endpoints` to `enpointsleases` first and later to leases.

So then why tests are failing here ?

**Comment by aojea** [comment]
```
E0203 13:21:18.190816   12050 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 47 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x150ae40, 0x247a6e0)
	/home/prow/go/pkg/mod/k8s.io/apimachinery@v0.19.1/pkg/util/runtime/runtime.go:74 +0xa3
```

The client-go vresion seems to old , 0.19.1, it has to bump it to a versions with the `endpointleases`

**Comment by NikhilSharmaWe** [comment]
@aojea Ok so someone else will going to bump the version (who is well informed about the repo) or we can do it ourselves?

**Comment by aojea** [comment]
you have to update in the go.mod, see an example

https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/commit/a4a426b2ac4744b2936640b0fdc529b9f11647a5

**Comment by NikhilSharmaWe** [comment]
@aojea
- I ran `go get k8s.io/apimachinery` in the repo. It update `go.mod` and `go.sum` file, still jobs are failing. What can be the reason ?

Apologies for asking time and time again.

**Comment by aojea** [comment]
check the job logs, it gives you hints

```
../../../pkg/mod/sigs.k8s.io/json@v0.0.0-20211020170558-c049b76a60c6/internal/golang/encoding/json/encode.go:1255:18: sf.IsExported undefined (type reflect.StructField has no field or method IsExported)
note: module requires Go 1.16
# golang.org/x/net/http2
../../../pkg/mod/golang.org/x/net@v0.0.0-20211209124913-491a49abca63/http2/transport.go:417:45: undefined: os.ErrDeadlineExceeded
note: module requires Go 1.17
```

I think that you can run those tests manually and test locally too

**Comment by NikhilSharmaWe** [comment]
@aojea I am not able to make any sense of these logs and know what is the cause of these jobs failing. Also you mentioned running these tests manually and locally, are you talking about running `test/e2e/test.sh` file locally.

**Comment by aojea** [comment]
I pasted the logs errorrs

> ../../../pkg/mod/golang.org/x/net@v0.0.0-20211209124913-491a49abca63/http2/transport.go:417:45: undefined: os.ErrDeadlineExceeded
note: module requires Go 1.17

it is failing because the golang dependencies are not correct

File: go.mod [hunk 1]
```diff
@@ -4,14 +4,12 @@ go 1.16
...
-	github.com/onsi/ginkgo v1.12.0 // indirect
-	github.com/onsi/gomega v1.9.0 // indirect
 	github.com/prometheus/client_golang v1.5.1
 	github.com/prometheus/client_model v0.2.0
 	golang.org/x/oauth2 v0.0.0-20200107190931-bf48bf16ab8d // indirect
 	golang.org/x/time v0.0.0-20191024005414-555d28b269f0
 	k8s.io/api v0.19.1
-	k8s.io/apimachinery v0.19.1
+	k8s.io/apimachinery v0.23.3
 	k8s.io/client-go v0.19.1
```
Comment (jsafrane) [inline go.mod:13]: you should bump this one too, to have all k8s libs from the same ver.

**Comment by aojea** [comment]
you are progressing, it fails now in 
```
sigs.k8s.io/sig-storage-lib-external-provisioner/v8/controller imports
	k8s.io/client-go/kubernetes imports
	k8s.io/client-go/rest imports
	k8s.io/client-go/plugin/pkg/client/auth/exec imports
	io/fs: malformed module path "io/fs": missing dot in first path element
make: *** [Makefile:18: dep] Error 1
```

**Comment by NikhilSharmaWe** [comment]
@aojea I am not sure about what to here.
As you mentioned this error is shown `io/fs: malformed module path "io/fs": missing dot in first path element`. So what we have to do here actually. 

Do we have to correct the module path manually somewhere or we have to import some dependencies to fix it ?

**Comment by aojea** [comment]
A quick search with that error in Google shows it requires a newer golang version ... I don't know where is that configured in this repo

File: go.mod (hunk 1 above)
Comment (NikhilSharmaWe) [reply go.mod]: @jsafrane As @aojea mentioned by seeing the [failing logs](https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/pull/120#issuecomment-1031926039) after bumping `k8s.io/client-go`, we need to new golang version for `io/fs` package but we don't have an idea where it is configured. 

Do you have any idea about this ? Please inform.

**Comment by jsafrane** [comment]
That's a good question, I *think* its here: https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes-sigs/sig-storage-lib-external-provisioner/sig-storage-lib-external-provisioner.yaml

**Comment by NikhilSharmaWe** [comment]
@aojea, @jsafrane
Changed the version of go image in [sig-storage-lib-external-provisioner.yaml](https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes-sigs/sig-storage-lib-external-provisioner/sig-storage-lib-external-provisioner.yaml) to a newer version, still jobs are failing.

But now they are not showing any logs, what could be the reason behind it ?

**Comment by NikhilSharmaWe** [comment]
@aojea Could you please give your thoughts on the previous [comment](https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/pull/120#issuecomment-1034042288).

**Comment by NikhilSharmaWe** [comment]
@aojea I have tried many times previously to build the repo locally but it seems like my system does not meet the minimum requirements. Is there any other way.

And what do you mean by `... because of that, it sasy go1.16 now`.

**Comment by aojea** [comment]
> @aojea I have tried many times previously to build the repo locally but it seems like my system does not meet the minimum requirements.

what requirements?

> And what do you mean by ... because of that, it sasy go1.16 now?



https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/a35bec6eb688ef7a0e31b340ed77bdb84010ace7/go.mod#L3

**Comment by jsafrane** [comment]
> it is just loading without showing any information about the fail.

The logs are too big, click on [Artifacts](https://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_sig-storage-lib-external-provisioner/120/pull-sig-storage-lib-external-provisioner-unit/1491469884767342592/) on the top right to get to a text file.

**Comment by NikhilSharmaWe** [comment]
@aojea Should I change the go version to `go 1.17` in go.mod file.

**Comment by NikhilSharmaWe** [comment]
@aojea Should I change the go version to go 1.17 in go.mod file.

**Comment by NikhilSharmaWe** [comment]
@aojea Should I change the go version to go 1.17 in go.mod file.

**Comment by aojea** [comment]
> @aojea Should I change the go version to go 1.17 in go.mod file.

I, honestl, can't guide you here,... if I were you I'll try to check locally what is the problem, if you can build and run unit tests locally

**Comment by BenTheElder** [comment]
the go version in `go.mod` specifies what go features should be available, you can have a `go 1.13` module built with go1.18, it just means you can't use generics and go1.13 should still work to build it.

but you need to have at least the version in go.mod installed on your machine.

**Comment by yonatankahana** [comment]
@NikhilSharmaWe do you plan to solve the failed tests soon? we really need to get this merged...

**Comment by humblec** [comment]
> @NikhilSharmaWe do you plan to solve the failed tests soon? we really need to get this merged...

Indeed... its important to get this in for consumers to use leader election APIs..  
@NikhilSharmaWe if no luck at your end, I can try to get a wrapper around this PR.. wdyt?

**Comment by NikhilSharmaWe** [comment]
@humblec Could you please help me what is the issue here. The script for gofmt is failing, I changed the code for it but still it is giving error.

**Comment by NikhilSharmaWe** [comment]
@yonatankahana Yes I also did that before but after that I was getting error related to that. Trying it again. Also do we need to change the go version in go.mod.

**Comment by yonatankahana** [comment]
> @yonatankahana Yes I also did that before but after that I was getting error related to that. Trying it again. Also do we need to change the go version in go.mod.

failing but from a different reason. the `repo-infra/verify/boilerplate/boilerplate.py` script still have 2 problems:
1. he searches for `python` which cannot be found, instead of `python3` (a guess, but probably correct)
2. he don't know about the `//go:build` syntax and it keep failing the test for both files you've fixed (enhance the regex: `go_build_constraints`)

**Comment by NikhilSharmaWe** [comment]
@yonatankahana If the the script `repo-infra/verify/boilerplate/boilerplate.py` don't know about the //go:build syntax then why are we adding it since the constraint is already added through // +build syntax. I believe it is due to when we upgrade the go version.

But I am not able to understand what type of tags are we searching for here. Could you please explain.
And apologies for spending too much time on this.

**Comment by yonatankahana** [comment]
> But I am not able to understand what type of tags are we searching for here. Could you please explain.
> And apologies for spending too much time on this.

@NikhilSharmaWe 
no need to apologize, it's team work :)

the `// +build` tags are deprecated and replaced by `//go:build` tags.
the script `boilerplate.py` verify that every file with code (.go, .py, Makefile, etc...) have proper [license header](https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/master/repo-infra/verify/boilerplate/boilerplate.go.txt).
to do that, as can be seen in [L84](https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/master/repo-infra/verify/boilerplate/boilerplate.py#L84) he wants to `remove build tags from the top of Go files`. to do that, he uses regex pattern that remove any line starts with `// +build`, but he don't know to remove lines starts with `//go:build` yet.
changing the regex in [L185](https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/master/repo-infra/verify/boilerplate/boilerplate.py#L185) from `^(// \+build.*\n)+\n` to `^(//(go:| \+)build.*\n)+\n` should detect both kinds and the test will pass. 
don't forget to edit the comment in [L184](https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/master/repo-infra/verify/boilerplate/boilerplate.py#L184) as well.

File: repo-infra/verify/boilerplate/boilerplate.py [hunk 2]
```diff
@@ -181,8 +181,8 @@ def get_regexs():
     # dates can be 2014, 2015, 2016, ..., CURRENT_YEAR, company holder names can be anything
     years = range(2014, date.today().year + 1)
     regexs["date"] = re.compile( '(%s)' % "|".join(map(lambda l: str(l), years)) )
-    # strip // +build \n\n build constraints
-    regexs["go_build_constraints"] = re.compile(r"^(// \+build.*\n)+\n", re.MULTILINE)
+    # strip (// go:build \n\n) and (// +build \n\n) build constraints
```
Comment (yonatankahana) [inline repo-infra/verify/boilerplate/boilerplate.py:184]: there should be no space between `//` and `go:build`

**Review by yonatankahana — CHANGES_REQUESTED** [review]

File: controller/controller.go [hunk 3]
```diff
@@ -865,7 +865,7 @@ func (ctrl *ProvisionController) Run(ctx context.Context) {
 	go ctrl.volumeStore.Run(ctx, DefaultThreadiness)
 
 	if ctrl.leaderElection {
-		rl, err := resourcelock.New("endpoints",
+		rl, err := resourcelock.New("endpointsleases",
```
Comment (yonatankahana) [inline controller/controller.go:868]: you can use the constant `resourcelock.EndpointsLeasesResourceLock` here

**Review by yonatankahana — APPROVED** [review]
i can confirm that i've checked it with `nfs-subdir-external-provisioner` and it looks good.

@wongma7, @jsafrane - can you take a look at it?

**Comment by humblec** [comment]
@yonatankahana++ for helping and getting this IN!!

**Pull request merged**